From ee481a02cb6ab946ec388efac7ca13d39305ff7b Mon Sep 17 00:00:00 2001
From: Frank Lichtenheld <frank@lichtenheld.com>
Date: Thu, 24 Apr 2025 10:35:14 +0200
Subject: [PATCH] mbedTLS: Always call mbedtls_ssl_set_hostname

This avoids new error
MBEDTLS_ERR_SSL_CERTIFICATE_VERIFICATION_WITHOUT_HOSTNAME
introduced in the latest mbedTLS 2.x and 3.x releases.
We usually do not use hostname verification via the TLS
library since we have other verification mechanisms.

Signed-off-by: Frank Lichtenheld <frank@lichtenheld.com>
---
 openvpn/mbedtls/ssl/sslctx.hpp | 13 +++++--------
 1 file changed, 5 insertions(+), 8 deletions(-)

diff --git a/openvpn/mbedtls/ssl/sslctx.hpp b/openvpn/mbedtls/ssl/sslctx.hpp
index a22adea48..8ccda8a9b 100644
--- a/openvpn/mbedtls/ssl/sslctx.hpp
+++ b/openvpn/mbedtls/ssl/sslctx.hpp
@@ -940,14 +940,11 @@ class MbedTLSContext : public SSLFactoryAPI
                     throw MbedTLSException("CA chain not defined");
 
                 // Set hostname for SNI or if a CA chain is configured
-                // In pre-mbedtls-2.x the hostname for the CA chain was set in ssl_set_ca_chain().
-                // From mbedtls-2.x, the hostname must be set via mbedtls_ssl_set_hostname()
-                // https://tls.mbed.org/kb/how-to/upgrade-2.0
-                if (hostname && ((c.flags & SSLConst::ENABLE_CLIENT_SNI) || c.ca_chain))
-                {
-                    if (mbedtls_ssl_set_hostname(ssl, hostname))
-                        throw MbedTLSException("mbedtls_ssl_set_hostname failed");
-                }
+                // Otherwise set the hostname explicitly to null to avoid
+                // MBEDTLS_ERR_SSL_CERTIFICATE_VERIFICATION_WITHOUT_HOSTNAME
+                bool use_hostname = hostname && ((c.flags & SSLConst::ENABLE_CLIENT_SNI) || c.ca_chain);
+                if (mbedtls_ssl_set_hostname(ssl, use_hostname ? hostname : nullptr))
+                    throw MbedTLSException("mbedtls_ssl_set_hostname failed");
 
                 // client cert+key
                 if (c.local_cert_enabled)
